dbms/src/Interpreters/Join.cpp

```diff
@@ -127,15 +123,27 @@ Join::Join(
             strictness = ASTTableJoin::Strictness::All;
         }
     }
+    if (!left_filter_column.empty() && !isLeftJoin(kind))
+        throw Exception("Not supported: non left join with left conditions");
+    if (!right_filter_column.empty() && !isRightJoin(kind))
+        throw Exception("Not supported: non right join with right conditions");
+}
+
+void Join::setBuildConcurrencyAndInitMap(size_t build_concurrency_)
+{
+    build_concurrency = std::max(1, build_concurrency_);
```

Comment:
assert `build_concurrency == 0 && pools.empty()`? how do you guarantee `setBuildConcurrencyAndInitMap` will and only be called once.